app/models/grade.rb

```diff
@@ -6,4 +6,16 @@ class Grade < ApplicationRecord
 
   belongs_to :exam
   belongs_to :enrollment, foreign_key: :enrollment_code, inverse_of: :grades
+
+  scope :average_grades_from_year, -> (year, id) {
+    joins(enrollment: :course, exam: :subject)
+    .where(enrollments: { student_id: id })
+    .where(courses: { year: })
+    .group('subjects.name', 'courses.name', 'courses.year', 'enrollments.code')
+    .select('subjects.name AS subject',
+            'courses.name AS course',
+            'enrollments.code AS code',
+            'AVG(grades.value) AS avg_grade')
+    .order('subjects.name')
+  }
```

Comment:
Since this query is complex, maybe this scope would look better as a class method,